src/amm.rs

```diff
@@ -45,25 +48,59 @@ macro_rules! to_u128 {
     };
 }
 
+macro_rules! multiply {
+    ($a:expr, $b:expr) => {
+        $a.checked_mul($b).expect(CANNOT_OVERFLOW)
```

Comment:
I think that these macros are not necessary. We basically wrapping only checked operations. 

I understand why you did. But that we have the error - we can probably completely remove the expectation that it "cannot_overflow" and return Overflow error instead. At the moment, with these types, it wont happen but it will be cleaner and safer imo ( so we dont start getting random panics in future but nice overflow errors ).

Reply:
Theres some beauty in it. Agreed.